test/sparse/bradc/sparseDomAssign2.good

```diff
@@ -40,7 +40,7 @@ A1 =
 0.0
 0.0
 0.0
-1.1
+5.5
```

Comment:
At first, I was surprised that this test's output would change.  But then looking at it a bit more, if I'm understanding properly, this has traditionally been incorrect (apparently without us realizing) and is now fixed?  What was the underlying cause?

Reply:
I had a feeling you were going to ask about this.  Traditionally it has been incorrect.  I am pretty sure the error was happening in the array preserving assignment.
```
// Test array-preserving assignment: (n/2, n/2) should be preserved
S1 = S2;
```
What method does that call so I can try to track down the changes that fixed it?

Reply:
It starts at the `  proc =(ref a: domain, b: domain) {` overload in ChapelArray.chpl, but the heavy lifting is done in `.dsiAssignDomain()` for the given sparse domain type.  For DefaultSparseDom, it looks like it's then implemented using either some local copies or `chpl_assignDomainWithIndsIterSafeForRemoving` which is implemented in ChapelDistribution.chpl and calls `dsiAdd()` and `dsiRemove`.

Reply:
Weird... the current behavior definitely seems more correct to me—like this should've been a bug all along.  But looking at dsiAdd() and dsiRemove() code, I'm not making the mental leap to why it would have changed...

Reply:
For those following along at home, we decided to not let this mystery hold up the merge since it's an improvement, but are still curious for the root cause.

Reply:
Mystery solved.  See https://github.com/chapel-lang/chapel/blob/48b5f16ebfcda1d117cfc1a7a46978a9a5c87831/modules/internal/DefaultSparse.chpl#L226-L228  
`a.sparseShiftArrayBack(insertPt.._nnz-1)` should have been `a.sparseShiftArrayBack(insertPt.._nnz)`.  Since it wasn't, the 5.5 value wasn't being moved into the position of being the first nonzero.  So the 1.1 value was the only nonzero at one point and then insertions of other nonzero indices (with the default IRV values) happened and 1.1 ended up where 5.5 should have been.  Here is some debugging output that shows what the old code was doing and then what the new code was doing.
```
// Output from code with bug, git checkout @{3.days.ago}
DEBUG: rem_help, before shift back, insertPt=3, _nnz-1=2
DEBUG in ChapelDistribtion: after shift back, data =1.1 2.2 3.3 4.4 5.5 6.6 7.7 8.8 9.9 0.0 0.0 0.0
DEBUG: rem_help, before shift back, insertPt=2, _nnz-1=1
DEBUG in ChapelDistribtion: after shift back, data =1.1 2.2 3.3 4.4 5.5 6.6 7.7 8.8 9.9 0.0 0.0 0.0
DEBUG: rem_help, before shift back, insertPt=1, _nnz-1=0
DEBUG in ChapelDistribtion: after shift back, data =1.1 2.2 3.3 4.4 5.5 6.6 7.7 8.8 9.9 0.0 0.0 0.0
DEBUG: _indices =(1, 9) (5, 5) (5, 5) (5, 5) (5, 5) (6, 6) (7, 7) (8, 8) (9, 9) (0, 0) (0, 0) (0, 0)
DEBUG: insertPt=1, _nnz-2=0, oldNNZDomSize=12, nnzDom.size-1=11
DEBUG in ChapelDistribtion: after shift, data =0.0 1.1 3.3 4.4 5.5 6.6 7.7 8.8 9.9 0.0 0.0 0.0
DEBUG: _indices =(1, 9) (2, 8) (5, 5) (5, 5) (5, 5) (6, 6) (7, 7) (8, 8) (9, 9) (0, 0) (0, 0) (0, 0)
DEBUG: insertPt=2, _nnz-2=1, oldNNZDomSize=12, nnzDom.size-1=11
DEBUG in ChapelDistribtion: after shift, data =0.0 0.0 1.1 4.4 5.5 6.6 7.7 8.8 9.9 0.0 0.0 0.0
DEBUG: _indices =(1, 9) (2, 8) (3, 7) (5, 5) (5, 5) (6, 6) (7, 7) (8, 8) (9, 9) (0, 0) (0, 0) (0, 0)
DEBUG: insertPt=3, _nnz-2=2, oldNNZDomSize=12, nnzDom.size-1=11
DEBUG in ChapelDistribtion: after shift, data =0.0 0.0 0.0 1.1 5.5 6.6 7.7 8.8 9.9 0.0 0.0 0.0
DEBUG: _indices =(1, 9) (2, 8) (3, 7) (4, 6) (5, 5) (6, 6) (7, 7) (8, 8) (9, 9) (0, 0) (0, 0) (0, 0)
DEBUG: insertPt=4, _nnz-2=3, oldNNZDomSize=12, nnzDom.size-1=11
DEBUG in ChapelDistribtion: after shift, data =0.0 0.0 0.0 0.0 1.1 6.6 7.7 8.8 9.9 0.0 0.0 0.0
DEBUG: _indices =(1, 9) (2, 8) (3, 7) (4, 6) (5, 5) (6, 4) (7, 7) (8, 8) (9, 9) (0, 0) (0, 0) (0, 0)
DEBUG: insertPt=6, _nnz-2=4, oldNNZDomSize=12, nnzDom.size-1=11
DEBUG in ChapelDistribtion: after shift, data =0.0 0.0 0.0 0.0 1.1 0.0 7.7 8.8 9.9 0.0 0.0 0.0
```


```
// Debug output from the current code
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, in lhs i=(5, 5)
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, in lhs i=(4, 4)
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, removing from lhs
DEBUG: rem_help, before shift back, insertPt=3, _nnz-1=3
DEBUG in ChapelDistribtion: after shift back, data =1.1 2.2 3.3 5.5 5.5 6.6 7.7 8.8 9.9 0.0 0.0 0.0
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, in lhs i=(3, 3)
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, removing from lhs
DEBUG: rem_help, before shift back, insertPt=2, _nnz-1=2
DEBUG in ChapelDistribtion: after shift back, data =1.1 2.2 5.5 5.5 5.5 6.6 7.7 8.8 9.9 0.0 0.0 0.0
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, in lhs i=(2, 2)
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, removing from lhs
DEBUG: rem_help, before shift back, insertPt=1, _nnz-1=1
DEBUG in ChapelDistribtion: after shift back, data =1.1 5.5 5.5 5.5 5.5 6.6 7.7 8.8 9.9 0.0 0.0 0.0
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, in lhs i=(1, 1)
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, removing from lhs
DEBUG: rem_help, before shift back, insertPt=0, _nnz-1=0
DEBUG in ChapelDistribtion: after shift back, data =5.5 5.5 5.5 5.5 5.5 6.6 7.7 8.8 9.9 0.0 0.0 0.0
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, in rhs i=(1, 9)
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, adding to lhs
DEBUG: _indices =(1, 9) (5, 5) (5, 5) (5, 5) (5, 5) (6, 6) (7, 7) (8, 8) (9, 9) (0, 0) (0, 0) (0, 0)
DEBUG: insertPt=0, _nnz-2=0, oldNNZDomSize=12, nnzDom.size-1=11
DEBUG in ChapelDistribtion: after shift, data =0.0 5.5 5.5 5.5 5.5 6.6 7.7 8.8 9.9 0.0 0.0 0.0
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, in rhs i=(2, 8)
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, adding to lhs
DEBUG: _indices =(1, 9) (2, 8) (5, 5) (5, 5) (5, 5) (6, 6) (7, 7) (8, 8) (9, 9) (0, 0) (0, 0) (0, 0)
DEBUG: insertPt=1, _nnz-2=1, oldNNZDomSize=12, nnzDom.size-1=11
DEBUG in ChapelDistribtion: after shift, data =0.0 0.0 5.5 5.5 5.5 6.6 7.7 8.8 9.9 0.0 0.0 0.0
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, in rhs i=(3, 7)
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, adding to lhs
DEBUG: _indices =(1, 9) (2, 8) (3, 7) (5, 5) (5, 5) (6, 6) (7, 7) (8, 8) (9, 9) (0, 0) (0, 0) (0, 0)
DEBUG: insertPt=2, _nnz-2=2, oldNNZDomSize=12, nnzDom.size-1=11
DEBUG in ChapelDistribtion: after shift, data =0.0 0.0 0.0 5.5 5.5 6.6 7.7 8.8 9.9 0.0 0.0 0.0
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, in rhs i=(4, 6)
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, adding to lhs
DEBUG: _indices =(1, 9) (2, 8) (3, 7) (4, 6) (5, 5) (6, 6) (7, 7) (8, 8) (9, 9) (0, 0) (0, 0) (0, 0)
DEBUG: insertPt=3, _nnz-2=3, oldNNZDomSize=12, nnzDom.size-1=11
DEBUG in ChapelDistribtion: after shift, data =0.0 0.0 0.0 0.0 5.5 6.6 7.7 8.8 9.9 0.0 0.0 0.0
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, in rhs i=(5, 5)
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, in rhs i=(6, 4)
DEBUG: chpl_assignDomainWithIndsIterSafeForRemoving, adding to lhs
DEBUG: _indices =(1, 9) (2, 8) (3, 7) (4, 6) (5, 5) (6, 4) (7, 7) (8, 8) (9, 9) (0, 0) (0, 0) (0, 0)
DEBUG: insertPt=5, _nnz-2=4, oldNNZDomSize=12, nnzDom.size-1=11
DEBUG in ChapelDistribtion: after shift, data =0.0 0.0 0.0 0.0 5.5 0.0 7.7 8.8 9.9 0.0 0.0 0.0
```

Reply:
Cool, thanks for figuring it out!